In [1]:
import pandas as pd
from sqlalchemy import create_engine, inspect, text
import datetime

# Database Connection

Connect to the PostgreSQL database with the provided credentials. SQLAlchemy will be used for database operations in this notebook.

In [2]:
# Connect to PostgreSQL using SQLAlchemy
try:
    engine = create_engine(
        'postgresql://azalea:azalea@localhost:5433/thelook_db'
    )
    # Test the connection
    with engine.connect() as conn:
        conn.execute(text('SELECT 1'))
    print("SQLAlchemy engine created successfully.")
except Exception as e:
    print(f"Failed to connect to database: {e}")
    print("\nPlease ensure that PostgreSQL database is running")
    raise

SQLAlchemy engine created successfully.


In [3]:
inspector = inspect(engine)
table_names = inspector.get_table_names()
table_names

['users',
 'events',
 'orders',
 'distribution_centers',
 'products',
 'inventory_items',
 'order_items']

In [4]:
# Read each table into a pandas DataFrame, assigning to individual variables
users = pd.read_sql_table('users', engine)
events = pd.read_sql_table('events', engine)
orders = pd.read_sql_table('orders', engine)
distribution_centers = pd.read_sql_table('distribution_centers', engine)
products = pd.read_sql_table('products', engine)
inventory_items = pd.read_sql_table('inventory_items', engine)
order_items = pd.read_sql_table('order_items', engine)

print('All tables read successfully.')

All tables read successfully.


In [5]:
# Close all database connections
try:
    if 'engine' in locals():
        engine.dispose()
        print("SQLAlchemy engine disposed.")
except Exception as e:
    print(f"Error disposing SQLAlchemy engine: {e}")

SQLAlchemy engine disposed.


# Exploratory Data Analysis

## Data Overview and Quality Assessment

Understanding the structure and quality of the data.

In [6]:
# Basic information about each dataset
print("Dataset Shapes:")
print(f"Users: {users.shape}")
print(f"Events: {events.shape}")
print(f"Orders: {orders.shape}")
print(f"Order Items: {order_items.shape}")
print(f"Products: {products.shape}")
print(f"Inventory Items: {inventory_items.shape}")
print(f"Distribution Centers: {distribution_centers.shape}")

# Memory usage
print("\nMemory Usage (MB):")
for name, df in [('users', users), ('events', events), ('orders', orders), 
                 ('order_items', order_items), ('products', products), 
                 ('inventory_items', inventory_items), ('distribution_centers', distribution_centers)]:
    print(f"{name}: {df.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

Dataset Shapes:
Users: (100000, 16)
Events: (2428216, 13)
Orders: (125278, 9)
Order Items: (181578, 11)
Products: (29120, 9)
Inventory Items: (490176, 12)
Distribution Centers: (10, 5)

Memory Usage (MB):
users: 67.77 MBusers: 67.77 MB
events: 1362.93 MB
orders: 19.50 MB
order_items: 23.77 MB
products: 10.62 MB
inventory_items: 189.79 MB
distribution_centers: 0.00 MB

events: 1362.93 MB
orders: 19.50 MB
order_items: 23.77 MB
products: 10.62 MB
inventory_items: 189.79 MB
distribution_centers: 0.00 MB


The datasets are large and detailed, especially `events`, which uses the most memory. Most tables have hundreds of thousands of rows, except for `distribution_centers`, which is small. Efficient filtering is important for analysis due to the size of the largest tables. Overall, the data is comprehensive and ready for business analytics.

In [7]:
# Data quality assessment
import numpy as np

def data_quality_summary(df, name):
    print(f"\n=== {name.upper()} DATA QUALITY ===")
    print(f"Shape: {df.shape}")
    print(f"Duplicates: {df.duplicated().sum()}")
    
    # Missing values
    missing = df.isnull().sum()
    missing_pct = (missing / len(df)) * 100
    missing_info = pd.DataFrame({
        'Missing Count': missing,
        'Missing %': missing_pct
    })
    missing_info = missing_info[missing_info['Missing Count'] > 0].sort_values('Missing %', ascending=False)
    
    if not missing_info.empty:
        print("\nMissing Values:")
        print(missing_info)
    else:
        print("\nNo missing values found!")
    
    # Data types
    print("\nData Types:")
    print(df.dtypes.value_counts())
    
    return missing_info

# Check each dataset
for name, df in [('users', users), ('orders', orders), ('order_items', order_items), 
                 ('products', products), ('inventory_items', inventory_items)]:
    data_quality_summary(df, name)


=== USERS DATA QUALITY ===
Shape: (100000, 16)
Duplicates: 0

No missing values found!

Data Types:
object                 11
int64                   2
float64                 2
datetime64[ns, UTC]     1
Name: count, dtype: int64

=== ORDERS DATA QUALITY ===
Shape: (125278, 9)
Duplicates: 0

Missing Values:
              Missing Count  Missing %
returned_at          112833  90.066093
delivered_at          81219  64.831016
shipped_at            43559  34.769872

Data Types:
datetime64[ns, UTC]    4
int64                  3
object                 2
Name: count, dtype: int64

=== ORDER_ITEMS DATA QUALITY ===
Shape: (181578, 11)
Duplicates: 0

No missing values found!

Data Types:
object                 11
int64                   2
float64                 2
datetime64[ns, UTC]     1
Name: count, dtype: int64

=== ORDERS DATA QUALITY ===
Shape: (125278, 9)
Duplicates: 0

Missing Values:
              Missing Count  Missing %
returned_at          112833  90.066093
delivered_at          8121

The data quality assessment shows:

- **Missing Values**: Minimal missing data across most tables. Geographic fields (latitude, longitude) and fulfillment dates (shipped_at, delivered_at) show expected gaps. Core business fields (IDs, prices, essential dates) are complete.

- **Data Types**: Appropriate data types throughout - numeric fields for calculations, text fields for categories, and timestamp fields for time series analysis.

- **Duplicates**: No duplicate records found, indicating proper data deduplication.

- **Overall**: The data quality is excellent for business analysis with minimal cleaning required. The completeness of critical fields enables comprehensive examination of customer behavior, product performance, and operational metrics.

Given the missing data patterns, we need to investigate whether these are truly missing values or represent natural business states (e.g., orders not yet shipped, products not yet sold).

In [8]:
# Missing Data Management and Analysis
print("=== MISSING DATA MANAGEMENT STRATEGY ===")
print("\n1. ORDERS TABLE - Analyzing fulfillment workflow missing data")

# Check if missing dates are due to order status rather than true missing data
order_status_analysis = orders.groupby('status').agg({
    'shipped_at': lambda x: x.isnull().sum(),
    'delivered_at': lambda x: x.isnull().sum(), 
    'returned_at': lambda x: x.isnull().sum()
}).astype(int)

order_status_analysis.columns = ['Missing_Shipped', 'Missing_Delivered', 'Missing_Returned']
order_status_analysis['Total_Orders'] = orders.groupby('status').size()

print("\nMissing fulfillment dates by order status:")
print(order_status_analysis)

# Calculate percentages
for col in ['Missing_Shipped', 'Missing_Delivered', 'Missing_Returned']:
    order_status_analysis[f'{col}_Pct'] = (order_status_analysis[col] / order_status_analysis['Total_Orders'] * 100).round(1)

print("\nMissing fulfillment dates by order status (with percentages):")
print(order_status_analysis[['Total_Orders', 'Missing_Shipped', 'Missing_Shipped_Pct', 
                           'Missing_Delivered', 'Missing_Delivered_Pct',
                           'Missing_Returned', 'Missing_Returned_Pct']])

=== MISSING DATA MANAGEMENT STRATEGY ===

1. ORDERS TABLE - Analyzing fulfillment workflow missing data

Missing fulfillment dates by order status:
            Missing_Shipped  Missing_Delivered  Missing_Returned  Total_Orders
status                                                                        
Cancelled             18714              18714             18714         18714
Complete                  0                  0             31614         31614
Processing            24845              24845             24845         24845
Returned                  0                  0                 0         12445
Shipped                   0              37660             37660         37660

Missing fulfillment dates by order status (with percentages):
            Total_Orders  Missing_Shipped  Missing_Shipped_Pct  \
status                                                           
Cancelled          18714            18714                100.0   
Complete           31614              

Analysis reveals that missing fulfillment dates represent the natural order processing workflow rather than data quality issues. Each status shows expected patterns:

- **Cancelled orders**: 100% missing all dates (never entered fulfillment)
- **Processing orders**: 100% missing all dates (awaiting shipment)
- **Shipped orders**: Have shipping dates but 100% missing delivered/returned dates
- **Complete orders**: No missing shipping/delivery dates, but 100% missing return dates
- **Returned orders**: Complete data across all date fields

These patterns confirm the database accurately tracks order lifecycle stages, with "missing" values actually serving as meaningful status indicators rather than data deficiencies.

In [9]:
print("2. ORDER_ITEMS TABLE - Analyzing item-level fulfillment data")

# Analyze missing data in order_items by status
order_items_analysis = order_items.groupby('status').agg({
    'shipped_at': lambda x: x.isnull().sum(),
    'delivered_at': lambda x: x.isnull().sum(),
    'returned_at': lambda x: x.isnull().sum()
}).astype(int)

order_items_analysis.columns = ['Missing_Shipped', 'Missing_Delivered', 'Missing_Returned']
order_items_analysis['Total_Items'] = order_items.groupby('status').size()

print("\nMissing fulfillment dates in order_items by status:")
print(order_items_analysis)

# Check if order_items missing data aligns with parent orders
print("\n\n3. INVENTORY_ITEMS TABLE - Analyzing sold_at missing data")

# For inventory_items, missing sold_at likely means items haven't been sold yet
inventory_sold_analysis = pd.DataFrame({
    'Total_Inventory_Items': [len(inventory_items)],
    'Items_with_sold_at': [inventory_items['sold_at'].notna().sum()],
    'Items_NOT_sold': [inventory_items['sold_at'].isnull().sum()],
    'Percentage_Unsold': [(inventory_items['sold_at'].isnull().sum() / len(inventory_items) * 100).round(1)]
})

print("\nInventory items sold status:")
print(inventory_sold_analysis)

# Calculate and print the sum of Items_with_sold_at and Items_NOT_sold
items_sold = inventory_sold_analysis['Items_with_sold_at'].iloc[0]
items_not_sold = inventory_sold_analysis['Items_NOT_sold'].iloc[0]
total_sum = items_sold + items_not_sold

print(f"\nItems with sold_at: {items_sold:,}")
print(f"Items NOT sold: {items_not_sold:,}")
print(f"Sum of both: {total_sum:,}")
print(f"Total inventory items (verification): {len(inventory_items):,}")
print(f"\nVerification: Sum equals total inventory items: {total_sum == len(inventory_items)}")

2. ORDER_ITEMS TABLE - Analyzing item-level fulfillment data

Missing fulfillment dates in order_items by status:
            Missing_Shipped  Missing_Delivered  Missing_Returned  Total_Items
status                                                                       
Cancelled             27190              27190             27190        27190
Complete                  0                  0             45955        45955
Processing            35839              35839             35839        35839
Returned                  0                  0                 0        17963
Shipped                   0              54631             54631        54631


3. INVENTORY_ITEMS TABLE - Analyzing sold_at missing data

Inventory items sold status:
   Total_Inventory_Items  Items_with_sold_at  Items_NOT_sold  \
0                 490176              181578          308598   

   Percentage_Unsold  
0               63.0  

Items with sold_at: 181,578
Items NOT sold: 308,598
Sum of both: 490,176
T

**Order Items Processing Flow**
- **Cancelled orders**: 100% missing all fulfillment dates (27,190 items)
- **Processing orders**: 100% missing all dates as they haven't been shipped (35,839 items)
- **Shipped orders**: Have shipping dates but await delivery (54,631 items)
- **Complete orders**: Full shipping and delivery information (45,955 items) 
- **Returned items**: Complete history across all date fields (17,963 items)

**Inventory Management Status**
- **37% of inventory sold**: 181,578 items have sold_at dates
- **63% of inventory unsold**: 308,598 items remain available in stock
- **Total inventory**: 490,176 items

These patterns indicate that missing date values represent meaningful business states in the order processing and inventory management workflows rather than data quality issues.

In [10]:
print("4. PRODUCTS TABLE - Analyzing missing product data")

# Analyze missing data patterns in products
products_missing = products.isnull().sum()
products_missing_pct = (products_missing / len(products) * 100).round(2)

products_missing_df = pd.DataFrame({
    'Missing_Count': products_missing,
    'Missing_Percentage': products_missing_pct
})
products_missing_df = products_missing_df[products_missing_df['Missing_Count'] > 0].sort_values('Missing_Percentage', ascending=False)

print("\nMissing data in products table:")
print(products_missing_df)

# Check if missing data correlates with certain categories or departments
if not products_missing_df.empty:
    print("\nAnalyzing patterns in missing product data:")
    
    # Check missing data by category
    for col in products_missing_df.index:
        if col in products.columns:
            missing_by_category = products.groupby('category')[col].apply(lambda x: x.isnull().sum())
            missing_by_category = missing_by_category[missing_by_category > 0].sort_values(ascending=False)
            
            if not missing_by_category.empty:
                print(f"\n{col} missing by category (top 5):")
                print(missing_by_category.head())
            
            # Check missing data by department
            missing_by_dept = products.groupby('department')[col].apply(lambda x: x.isnull().sum())
            missing_by_dept = missing_by_dept[missing_by_dept > 0].sort_values(ascending=False)
            
            if not missing_by_dept.empty:
                print(f"\n{col} missing by department:")
                print(missing_by_dept)

4. PRODUCTS TABLE - Analyzing missing product data

Missing data in products table:
       Missing_Count  Missing_Percentage
brand             24                0.08
name               2                0.01

Analyzing patterns in missing product data:

brand missing by category (top 5):
category
Intimates            4
Tops & Tees          4
Outerwear & Coats    3
Swim                 2
Accessories          2
Name: brand, dtype: int64

brand missing by department:
department
Men      12
Women    12
Name: brand, dtype: int64

name missing by category (top 5):
category
Intimates            1
Outerwear & Coats    1
Name: name, dtype: int64

name missing by department:
department
Men      1
Women    1
Name: name, dtype: int64


The products table shows excellent data quality with very limited missing information. The few missing values are evenly distributed across departments, suggesting random omissions rather than systematic data issues. table shows excellent data quality with very limited missing information. The few missing values are evenly distributed across departments, suggesting random omissions rather than systematic data issues.

In [11]:
# Date range analysis
print("=== DATE RANGES ===")

# Convert date columns to datetime if they aren't already
date_columns = {
    'users': ['created_at'],
    'events': ['created_at'],
    'orders': ['created_at', 'returned_at', 'shipped_at', 'delivered_at'],
    'order_items': ['created_at', 'shipped_at', 'delivered_at', 'returned_at'],
    'inventory_items': ['created_at', 'sold_at']
}

# Track sources of min/max dates
min_date_source = None
max_date_source = None

def convert_and_range(df, col, table_name):
    global overall_min, overall_max, min_date_source, max_date_source
    
    if col in df.columns:
        if not pd.api.types.is_datetime64_any_dtype(df[col]):
            df[col] = pd.to_datetime(df[col], errors='coerce')
        valid_dates = df[col].dropna()
        if not valid_dates.empty:
            min_date, max_date = valid_dates.min(), valid_dates.max()
            print(f"  {col}: {min_date.strftime('%d %b %Y')} to {max_date.strftime('%d %b %Y')}")
            print(f"    Range: {(max_date - min_date).days} days")
            
            # Update overall min date if this is earlier
            if overall_min is None or min_date < overall_min:
                overall_min = min_date
                min_date_source = f"{table_name}.{col}"
                
            # Update overall max date if this is later
            if overall_max is None or max_date > overall_max:
                overall_max = max_date
                max_date_source = f"{table_name}.{col}"
                
            return min_date, max_date
        else:
            print(f"  {col}: No valid dates")
    return None, None

overall_min, overall_max = None, None
for table, cols in date_columns.items():
    df = locals()[table]
    print(f"\n{table.upper()}:")
    for col in cols:
        min_date, max_date = convert_and_range(df, col, table)

if overall_min and overall_max:
    print("\n=== OVERALL DATE RANGE ===")
    print(f"\nOldest date: {overall_min.strftime('%d %b %Y')} (from {min_date_source})")
    print(f"Newest date: {overall_max.strftime('%d %b %Y')} (from {max_date_source})")
    print(f"Total time span: {(overall_max - overall_min).days} days")

=== DATE RANGES ===

USERS:
  created_at: 02 Jan 2019 to 24 May 2025
    Range: 2334 days

EVENTS:
  created_at: 02 Jan 2019 to 29 May 2025
    Range: 2339 days

ORDERS:
  created_at: 06 Jan 2019 to 25 May 2025
    Range: 2331 days
  returned_at: 25 Jan 2019 to 04 Jun 2025
    Range: 2321 days
  shipped_at: 06 Jan 2019 to 28 May 2025
    Range: 2334 days
  delivered_at: 13 Jan 2019 to 02 Jun 2025
    Range: 2331 days

ORDER_ITEMS:
  created_at: 06 Jan 2019 to 29 May 2025
    Range: 2335 days
  shipped_at: 06 Jan 2019 to 28 May 2025
    Range: 2334 days
  delivered_at: 13 Jan 2019 to 02 Jun 2025
    Range: 2331 days
  returned_at: 25 Jan 2019 to 04 Jun 2025
    Range: 2321 days

INVENTORY_ITEMS:
  created_at: 21 Nov 2018 to 29 May 2025
    Range: 2380 days
  sold_at: 06 Jan 2019 to 29 May 2025
    Range: 2335 days

=== OVERALL DATE RANGE ===

Oldest date: 21 Nov 2018 (from inventory_items.created_at)
Newest date: 04 Jun 2025 (from orders.returned_at)
Total time span: 2386 days
  deliver

The dataset spans from `2018-11-21` (`inventory_items.created_at`) to `2025-06-04` (`orders.returned_at`), covering 2,386 days. Key date columns like `users.created_at`, `events.created_at`, and `orders.created_at` show continuous records from early 2019 to mid-2025, indicating sufficient data for longitudinal analysis.

## Business and Performance Metrics Overview

In [12]:
print("=== KEY BUSINESS METRICS ===")

# Basic counts
total_users = users['id'].nunique()
total_orders = orders['order_id'].nunique()
total_products = products['id'].nunique()
total_revenue = order_items['sale_price'].sum()

print(f"Total Users: {total_users:,}")
print(f"Total Orders: {total_orders:,}")
print(f"Total Products: {total_products:,}")
print(f"Total Revenue: ${total_revenue:,.2f}")

# Calculate key metrics
users_with_orders = orders['user_id'].nunique()
conversion_rate = (users_with_orders / total_users) * 100
aov = total_revenue / total_orders
median_order_value = order_items.groupby('order_id')['sale_price'].sum().median()
revenue_per_user = total_revenue / total_users
revenue_per_customer = total_revenue / users_with_orders

print(f"\n=== PERFORMANCE METRICS ===")
print(f"Users with Orders: {users_with_orders:,}")
print(f"Conversion Rate: {conversion_rate:.2f}%")
print(f"Average Order Value: ${aov:.2f}")
print(f"Median Order Value: ${median_order_value:.2f}")
print(f"Revenue per User: ${revenue_per_user:.2f}")
print(f"Revenue per Customer: ${revenue_per_customer:.2f}")

# Order statistics
orders_per_customer = orders.groupby('user_id').size()
repeat_customers = (orders_per_customer > 1).sum()
repeat_customer_rate = (repeat_customers / users_with_orders) * 100
returned_orders = orders[orders['status'].str.lower().str.contains('return')]['order_id'].nunique()
order_return_rate = (returned_orders / total_orders) * 100 if total_orders else 0
items_per_order = order_items.groupby('order_id').size().mean()
cutoff_date = (pd.Timestamp(datetime.date.today()) - pd.DateOffset(months=6)).tz_localize('UTC')
last_order_dates = orders.groupby('user_id')['created_at'].max()
churned_users = (last_order_dates < cutoff_date).sum()
churn_rate = (churned_users / users_with_orders) * 100 if users_with_orders else 0
print(f"\n=== ORDER PATTERNS ===")
print(f"Avg Orders per Customer: {orders_per_customer.mean():.2f}")
print(f"Max Orders by Single Customer: {orders_per_customer.max()}")
print(f"% of One-time Customers: {(orders_per_customer == 1).mean() * 100:.2f}%")
print(f"Repeat Customer Rate: {repeat_customer_rate:.2f}%")
print(f"Order Return Rate: {order_return_rate:.2f}%")
print(f"Average Items per Order: {items_per_order:.2f}")
print(f"Churn Rate (no order in last 6 months): {churn_rate:.2f}%")

=== KEY BUSINESS METRICS ===
Total Users: 100,000
Total Orders: 125,278
Total Products: 29,120
Total Revenue: $10,788,195.71

=== PERFORMANCE METRICS ===
Users with Orders: 79,989
Conversion Rate: 79.99%
Average Order Value: $86.11
Median Order Value: $55.03
Revenue per User: $107.88
Revenue per Customer: $134.87

=== PERFORMANCE METRICS ===
Users with Orders: 79,989
Conversion Rate: 79.99%
Average Order Value: $86.11
Median Order Value: $55.03
Revenue per User: $107.88
Revenue per Customer: $134.87

=== ORDER PATTERNS ===
Avg Orders per Customer: 1.57
Max Orders by Single Customer: 4
% of One-time Customers: 62.25%
Repeat Customer Rate: 37.75%
Order Return Rate: 9.93%
Average Items per Order: 1.45
Churn Rate (no order in last 6 months): 65.02%

=== ORDER PATTERNS ===
Avg Orders per Customer: 1.57
Max Orders by Single Customer: 4
% of One-time Customers: 62.25%
Repeat Customer Rate: 37.75%
Order Return Rate: 9.93%
Average Items per Order: 1.45
Churn Rate (no order in last 6 months): 65

In [13]:
# Other performance metrics

# Average Shipping and Delivery Time (in days)
orders_shipped = orders[orders['shipped_at'].notna() & orders['created_at'].notna()]
orders_delivered = orders[orders['delivered_at'].notna() & orders['shipped_at'].notna()]
if not orders_shipped.empty:
    avg_ship_time = (orders_shipped['shipped_at'] - orders_shipped['created_at']).dt.total_seconds().mean() / 86400
    print(f"Average Time to Ship: {avg_ship_time:.2f} days")
else:
    print("\nAverage Time to Ship: N/A")
if not orders_delivered.empty:
    avg_delivery_time = (orders_delivered['delivered_at'] - orders_delivered['shipped_at']).dt.total_seconds().mean() / 86400
    print(f"Average Time to Deliver: {avg_delivery_time:.2f} days")
else:
    print("Average Time to Deliver: N/A")
    
# Customer Segmentation: Repeat vs. One-time Customers
one_time = (orders_per_customer == 1).sum()
repeat = (orders_per_customer > 1).sum()
print(f"\nCustomer Segmentation:")
print(f"  One-time Customers: {one_time:,} ({(one_time/total_users)*100:.2f}%)")
print(f"  Repeat Customers: {repeat:,} ({(repeat/total_users)*100:.2f}%)")
    
# Conversion Rate by Traffic Source
traffic_conv = users.merge(orders[['user_id']], left_on='id', right_on='user_id', how='left', indicator=True)
traffic_conv['is_buyer'] = traffic_conv['_merge'] == 'both'
traffic_conv_deduped = traffic_conv.drop_duplicates('id')
conv_by_source = traffic_conv_deduped.groupby('traffic_source').agg(
    total_users=('id','count'),
    buyers=('is_buyer','sum')
)
conv_by_source['conversion_rate'] = (conv_by_source['buyers'] / conv_by_source['total_users']) * 100
print("\nConversion Rate by Traffic Source:")
for idx, row in conv_by_source.iterrows():
    print(f"  {idx}: {row['conversion_rate']:.2f}% ({int(row['buyers'])}/{int(row['total_users'])})")
    
# Order status distribution
status_counts = orders['status'].value_counts()
print("\nOrder Status Distribution:")
for status, count in status_counts.items():
    pct = (count / total_orders) * 100
    print(f"  {status}: {count:,} orders ({pct:.1f}%)")

Average Time to Ship: 1.50 days
Average Time to Deliver: 2.49 days

Customer Segmentation:
  One-time Customers: 49,795 (49.80%)
  Repeat Customers: 30,194 (30.19%)

Conversion Rate by Traffic Source:
  Display: 79.69% (3189/4002)
  Email: 80.37% (3917/4874)
  Facebook: 79.90% (4930/6170)
  Organic: 80.16% (11883/14824)
  Search: 79.95% (56070/70130)

Order Status Distribution:
  Shipped: 37,660 orders (30.1%)
  Complete: 31,614 orders (25.2%)
  Processing: 24,845 orders (19.8%)
  Cancelled: 18,714 orders (14.9%)
  Returned: 12,445 orders (9.9%)

Conversion Rate by Traffic Source:
  Display: 79.69% (3189/4002)
  Email: 80.37% (3917/4874)
  Facebook: 79.90% (4930/6170)
  Organic: 80.16% (11883/14824)
  Search: 79.95% (56070/70130)

Order Status Distribution:
  Shipped: 37,660 orders (30.1%)
  Complete: 31,614 orders (25.2%)
  Processing: 24,845 orders (19.8%)
  Cancelled: 18,714 orders (14.9%)
  Returned: 12,445 orders (9.9%)


## Top Customers, Categories, Brands, and Products by Revenue and Popularity

In [14]:
# Top 5 Customer Lifetime Value (CLV)
customer_revenue = orders.merge(order_items, left_on='order_id', right_on='order_id')
grouped_customers = customer_revenue.groupby('user_id_x')['sale_price'].sum().sort_values(ascending=False)
top_customers = users.set_index('id').loc[grouped_customers.head(5).index]
top_customers = top_customers[['first_name', 'last_name']].copy()
top_customers['revenue'] = grouped_customers.head(5).values
print("Top 5 Customers by Lifetime Value (CLV):")
for i, row in top_customers.iterrows():
    print(f"  {row['first_name']} {row['last_name']}: ${row['revenue']:,.2f}")

# Top 5 category by revenue
category_revenue = order_items.merge(products[['id','category']], left_on='product_id', right_on='id')
category_revenue = category_revenue.groupby('category')['sale_price'].sum().sort_values(ascending=False)
print("\nTop 5 Product Categories by Revenue:")
for cat, val in category_revenue.head(5).items():
    print(f"  {cat}: ${val:,.2f}")

# Top 5 category by order count (popularity)
category_popularity = order_items.merge(products[['id','category']], left_on='product_id', right_on='id')
category_order_count = category_popularity.groupby('category')['order_id'].nunique().sort_values(ascending=False)
print("\nTop 5 Product Categories by Order Count:")
for cat, count in category_order_count.head(5).items():
    print(f"  {cat}: {count:,} orders")

# Top 5 brand by revenue
brand_revenue = order_items.merge(products[['id','brand']], left_on='product_id', right_on='id')
brand_revenue = brand_revenue.groupby('brand')['sale_price'].sum().sort_values(ascending=False)
print("\nTop 5 Brands by Revenue:")
for brand, val in brand_revenue.head(5).items():
    print(f"  {brand}: ${val:,.2f}")

# Top 5 brand by order count (popularity)
brand_popularity = order_items.merge(products[['id','brand']], left_on='product_id', right_on='id')
brand_order_count = brand_popularity.groupby('brand')['order_id'].nunique().sort_values(ascending=False)
print("\nTop 5 Brands by Order Count:")
for brand, count in brand_order_count.head(5).items():
    print(f"  {brand}: {count:,} orders")

# Top 5 products by revenue
product_revenue = order_items.groupby('product_id')['sale_price'].sum().sort_values(ascending=False)
top_products = products.set_index('id').loc[product_revenue.head(5).index]
top_products = top_products[['name', 'retail_price']].copy()
top_products['revenue'] = product_revenue.head(5).values
print("\nTop 5 Products by Revenue:")
for i, row in top_products.iterrows():
    print(f"  {row['name']} (Price: ${row['retail_price']:.2f}): ${row['revenue']:,.2f}")

# Top 5 products by order count (popularity)
product_order_count = order_items['product_id'].value_counts().head(5)
top_products_count = products.set_index('id').loc[product_order_count.index]
top_products_count = top_products_count[['name', 'retail_price']].copy()
top_products_count['order_count'] = product_order_count.values
print("\nTop 5 Products by Order Amount:")
for i, row in top_products_count.iterrows():
    print(f"  {row['name']} (Price: ${row['retail_price']:.2f}): {row['order_count']:,} orders")

Top 5 Customers by Lifetime Value (CLV):
  Charles Spencer: $2,015.92
  Devin White: $1,626.13
  Robert Gonzalez: $1,535.02
  Joseph Stewart: $1,520.93
  Jennifer Ramos: $1,490.99

Top 5 Product Categories by Revenue:
  Outerwear & Coats: $1,291,409.80
  Jeans: $1,255,639.76
  Sweaters: $830,785.40
  Swim: $647,627.43
  Suits & Sport Coats: $647,153.19

Top 5 Product Categories by Order Count:
  Intimates: 12,713 orders
  Jeans: 12,434 orders
  Tops & Tees: 11,564 orders
  Fashion Hoodies & Sweatshirts: 11,344 orders
  Swim: 11,076 orders

Top 5 Brands by Revenue:
  Diesel: $201,557.41
  Calvin Klein: $201,503.74
  True Religion: $179,670.11
  7 For All Mankind: $172,409.43
  Carhartt: $169,504.77

Top 5 Brands by Order Count:
  Allegra K: 6,130 orders
  Calvin Klein: 3,112 orders
  Carhartt: 2,506 orders
  Hanes: 1,956 orders
  Volcom: 1,850 orders

Top 5 Products by Revenue:
  The North Face Apex Bionic Soft Shell Jacket - Men's (Price: $903.00): $11,739.00
  ASCIS Cushion Low Socks 

## Customer Demography

In [16]:
# --- Customer Demography EDA ---
print("=== CUSTOMER DEMOGRAPHY ===")

# Gender distribution
if 'gender' in users.columns:
    gender_counts = users['gender'].value_counts(dropna=False)
    gender_pct = users['gender'].value_counts(normalize=True, dropna=False) * 100
    print("\nGender Distribution:")
    for g, c in gender_counts.items():
        print(f"  {g}: {c:,} ({gender_pct[g]:.2f}%)")
else:
    print("\nNo gender data available.")

# Age distribution
if 'age' in users.columns:
    print("\nAge Statistics:")
    print(f"  Mean Age: {users['age'].mean():.1f}")
    print(f"  Median Age: {users['age'].median():.1f}")
    print(f"  Min Age: {users['age'].min()}")
    print(f"  Max Age: {users['age'].max()}")
    print("  Age Groups:")
    age_bins = [0, 18, 25, 35, 45, 55, 65, 100]
    age_labels = ['<18', '18-24', '25-34', '35-44', '45-54', '55-64', '65+']
    users['age_group'] = pd.cut(users['age'], bins=age_bins, labels=age_labels, right=False)
    age_group_counts = users['age_group'].value_counts(sort=False)
    for group, count in age_group_counts.items():
        pct = (count / len(users)) * 100
        print(f"    {group}: {count:,} ({pct:.2f}%)")
else:
    print("\nNo age data available.")

# Geographic distribution (country, state, city)
for col in ['country', 'state', 'city']:
    if col in users.columns:
        top_geo = users[col].value_counts().head(5)
        print(f"\nTop 5 {col.title()}s by User Count:")
        for val, cnt in top_geo.items():
            print(f"  {val}: {cnt:,}")
    else:
        print(f"\nNo {col} data available.")

# Traffic source breakdown
if 'traffic_source' in users.columns:
    traffic_counts = users['traffic_source'].value_counts()
    traffic_pct = users['traffic_source'].value_counts(normalize=True) * 100
    print("\nTraffic Source Distribution:")
    for src, cnt in traffic_counts.items():
        print(f"  {src}: {cnt:,} ({traffic_pct[src]:.2f}%)")
else:
    print("\nNo traffic source data available.")

# Customer tenure (how long since registration)
if 'created_at' in users.columns:
    users['created_at'] = pd.to_datetime(users['created_at'], errors='coerce')
    # Make the current timestamp timezone-aware with UTC to match users['created_at']
    today = pd.Timestamp(datetime.date.today(), tz='UTC')
    users['tenure_days'] = (today - users['created_at']).dt.days
    print("\nCustomer Tenure (days since registration):")
    print(f"  Mean: {users['tenure_days'].mean():.1f}")
    print(f"  Median: {users['tenure_days'].median():.1f}")
    print(f"  Min: {users['tenure_days'].min()}")
    print(f"  Max: {users['tenure_days'].max()}")
else:
    print("\nNo registration date data available.")

# Customer segmentation by demography and purchasing
if 'gender' in users.columns and 'age_group' in users.columns:
    # Merge with orders to get purchasing info
    user_orders = orders.groupby('user_id').size().rename('order_count')
    users_demo = users.join(user_orders, how='left').fillna({'order_count': 0})
    print("\nAverage Orders per User by Gender and Age Group:")
    seg = users_demo.groupby(['gender', 'age_group'], observed=True)['order_count'].mean().unstack()
    print(seg.round(2))
else:
    print("\nNot enough demographic data for segmentation.")

# Top cities/states/countries by revenue (if location and orders can be joined)
if 'city' in users.columns:
    user_revenue = orders.merge(users[['id', 'city', 'state', 'country']], left_on='user_id', right_on='id', how='left')
    city_rev = user_revenue.groupby('city')['order_id'].count().sort_values(ascending=False).head(5)
    print("\nTop 5 Cities by Order Count:")
    for city, cnt in city_rev.items():
        print(f"  {city}: {cnt:,} orders")
else:
    print("\nNo city data for revenue analysis.")

=== CUSTOMER DEMOGRAPHY ===

Gender Distribution:
  F: 50,172 (50.17%)
  M: 49,828 (49.83%)

Age Statistics:
  Mean Age: 41.0
  Median Age: 41.0
  Min Age: 12
  Max Age: 70
  Age Groups:
    <18: 10,195 (10.20%)
    18-24: 11,756 (11.76%)
    25-34: 16,976 (16.98%)
    35-44: 16,828 (16.83%)
    45-54: 16,980 (16.98%)
    55-64: 17,152 (17.15%)
    65+: 10,113 (10.11%)

Top 5 Countrys by User Count:
  China: 33,821
  United States: 22,516
  Brasil: 14,511
  South Korea: 5,304
  United Kingdom: 4,711

Top 5 States by User Count:
  Guangdong: 5,425
  England: 4,170
  California: 3,745
  Texas: 2,513
  Shanghai: 2,423

Top 5 Citys by User Count:
  Shanghai: 2,483
  Beijing: 2,042
  Seoul: 1,444
  Shenzhen: 1,292
  null: 992

Traffic Source Distribution:
  Search: 70,130 (70.13%)
  Organic: 14,824 (14.82%)
  Facebook: 6,170 (6.17%)
  Email: 4,874 (4.87%)
  Display: 4,002 (4.00%)

Customer Tenure (days since registration):
  Mean: 1144.6
  Median: 1143.0
  Min: 7
  Max: 2341

Average Orders